bittensor/_synapse/text_causallmnext_impl.py

```diff
@@ -0,0 +1,170 @@
+            forward_response_serializer_type=self.forward_response_serializer_type,
+            backward_request_serializer_type=self.backward_request_serializer_type,
+            backward_response_serializer_type=self.backward_response_serializer_type,
+        )
+
+    def serialize_to_wire_proto(self, code: 'bittensor.proto.ReturnCode' = 0,
+                                message: str = '') -> bittensor.proto.Synapse:
+        return bittensor.proto.Synapse(
+            synapse_data=self.serialize_to_instance_proto().SerializeToString(),
+            synapse_type=TextCausalLMNext.synapse_type,
+            return_code=code,
+            message=message
+        )
+
+    def check_forward_request_tensor(self, forward_request_tensor):
+        # forward_request_tensor: [batch_size, sequence_len]
+        if (
+                len(forward_request_tensor.shape) != 2 or
+                forward_request_tensor.shape[0] == 0 or
+                forward_request_tensor.shape[1] == 0
+        ):
+            raise ValueError(f"forward_request_tensor.shape must be in [-1, -1], "
+                             f"got: {list(forward_request_tensor.shape)} for synapse: {self}")
+
+    def check_forward_response_tensor(self, forward_request_tensor, forward_response_tensor):
+        # forward_request_tensor: [batch_size, sequence_len]
+        # forward_response_tensor: [ >= batch_size * (2 * topk + 1)]
+        if forward_response_tensor is None:
+            raise ValueError("Empty Response")
+
+        if (
+                len(forward_response_tensor.shape) != 1 or
+                forward_response_tensor.size(0) < forward_request_tensor.shape[0] * (2 * self.topk + 1)
+        ):
+            raise ValueError(f"forward_response_tensor.shape must be in "
+                             f"[>={forward_request_tensor.shape[0]} x (2 x {self.topk} + 1)], "
+                             f"got: {forward_response_tensor.size(0)} for synapse: {self}")
+
+    def check_backward_request_gradient(self, forward_request_tensor, backward_request_gradient):
+        # forward_request_tensor: [batch_size, sequence_len]
+        # backward_request_gradient: [ >= batch_size * (2 * topk + 1)]
+        if (
+                len(backward_request_gradient.shape) != 1 or
+                backward_request_gradient.size(0) < forward_request_tensor.shape[0] * (2 * self.topk + 1)
+        ):
+            raise ValueError(f"backward_request_gradient.shape must be in "
+                             f"[>={forward_request_tensor.shape[0]} x (2 x {self.topk} + 1)], "
+                             f"got: {backward_request_gradient.size(0)} for synapse: {self}")
+
+    def encode_forward_request_tensor(self, forward_request_tensor: torch.Tensor) -> torch.Tensor:
+        return forward_request_tensor
+
+    def decode_forward_request_tensor(self, forward_request_tensor: torch.Tensor) -> torch.Tensor:
+        return forward_request_tensor
```

Comment:
Do we need encoding or decoding function?

Reply:
Currently, encoding and decoding is best performed at the origin, that is server and validator forward() functions respectively. The reason being that variable length phrases need to otherwise be converted to padded tensors, simply for the purpose of passing it to the synapse encode_forward_request_tensor(), where it would need to be decoded and then re-encoded and compacted, thus duplicating work best done immediately at the origin.